Delete hm port on unit removal

zhhuabj opened the pull request:
This is Zaza functional test for lp:1915512. Remove a unit, and then query
the list of ports to check that the port has been deleted.

Depends-On: https://review.opendev.org/c/openstack/charm-octavia/+/787700

AurelienLourot commented on zaza/openstack/charm_tests/octavia/tests.py:
```diff
@@ -99,11 +100,77 @@ def revoke_role_current_user(keystone_client, keystone_session, role_name,
+        zaza.model.add_unit(self.application_name)
+        logging.info("Wait until one unit is added ...")
+        zaza.model.block_until_unit_count(self.application_name, num+1)
+        zaza.model.wait_for_application_states(states=self.states)
```
This is the same as `zaza.model.wait_for_application_states()`, so you don't need `self.states` at all. This is a bit lengthy to explain here, but in a nutshell, both will lead to "waiting for **all** applications to be active/ready"

zhhuabj replied:
done

AurelienLourot commented on zaza/openstack/charm_tests/octavia/tests.py:
```diff
@@ -99,11 +100,77 @@ def revoke_role_current_user(keystone_client, keystone_session, role_name,
     """Charm operation tests."""
 
+    application_name = "octavia"
+    states = {"octavia": {
```
You don't need this, see other comment, thanks!

zhhuabj replied:
done

AurelienLourot commented on zaza/openstack/charm_tests/octavia/tests.py:
```diff
@@ -99,11 +100,77 @@ def revoke_role_current_user(keystone_client, keystone_session, role_name,
 class CharmOperationTest(test_utils.OpenStackBaseTest):
     """Charm operation tests."""
 
+    application_name = "octavia"
```
You don't need this, see other comment, thanks!

zhhuabj replied:
done

AurelienLourot commented on zaza/openstack/charm_tests/octavia/tests.py:
```diff
@@ -99,11 +100,77 @@ def revoke_role_current_user(keystone_client, keystone_session, role_name,
 
+    def get_leaders_and_non_leaders(self):
+        """Get leader node and non-leader nodes of octavia."""
+        status = zaza.model.get_status().applications[self.application_name]
```
Instead of `self.application_name` you can use `self.test_config['charm_name']`

AurelienLourot commented on zaza/openstack/charm_tests/octavia/tests.py:
```diff
@@ -99,11 +100,77 @@ def revoke_role_current_user(keystone_client, keystone_session, role_name,
         """Run class setup for running Octavia charm operation tests."""
         super(CharmOperationTest, cls).setUpClass()
 
+    def get_leaders_and_non_leaders(self):
```
You copy-pasted this function from `mysql/tests.py`. Please do not duplicate code. Move it to a central place instead. Thanks!

zhhuabj replied:
move to zaza/openstack/utilities/generic.py now

AurelienLourot commented on zaza/openstack/charm_tests/octavia/tests.py:
```diff
@@ -99,11 +100,77 @@ def revoke_role_current_user(keystone_client, keystone_session, role_name,
+    def test_update_controller_ip_port_list(self):
+        """Test update_controller_ip_port_list.
+
+        Remove a unit, and then query the list of ports to check that the port
```
It's actually adding a unit and then deleting a unit

AurelienLourot reviewed (changes requested):
This looks good in general. See my inline comments, thanks! Also, since you are a first-time contributer: from now on please don't push-force, or it will make it difficult for me to re-review. Thanks!

zhhuabj commented:
Hi Aurelien,

I have addressed all your comment except 'don not push-force, or it will make it difficult for me to re-review'.

Because if I don't use '-f', then I will see the error '! [rejected]        hm-port -> hm-port (non-fast-forward)', see https://paste.ubuntu.com/p/yjxRFd4ZkF/

First, my understanding is that using '-f' will not override any code, because I didn't submit the code into upstream's master branch directly, I just submitted it into my own repo and also used another branch name. After PR is approved, only commits are merged into upstream's master. So using '-f' should not wash out the upstream newly commits.

Here are my steps to update PR:

# update local master branch from upstream
git checkout master
git pull --rebase upstream master

# use rebase to update local hm-port branch from upstream
git checkout hm-port
git rebase upstream/master hm-port

# commit the code with '--amend'
git commit --amend

# git push the code with '-f', if I don't use '-f', I will hit above rejected error
git push origin hm-port -f

I searched the meaning of '-f', I got:

The -f is actually required because of the rebase. Whenever you do a rebase you would need to do a force push because the remote branch cannot be fast-forwarded to your commit. You'd always want to make sure that you do a pull before pushing, but if you don't like to force push to master or dev for that matter, you can create a new branch to push to and then merge or make a PR.

So it looks '-f' is required, right? and I also found another two patches [2][3], they are also using '-f'.

I always use 'git commit --amend' to commit the code, using '--amend' means hashid is different, so it seems to make sense to use '-f' as well. When using '--amend', gerrit will show you the difference of each submission, so it's convenient for you to review. But for RP, I don't know how to achieve this point. Don't we need to use '--amend', but that will generate many commits, it doesn't seem to be what we want. We may also need '--amend'.

Sorry, I might not understand the meaning of your that sentence, can you clarify that? or let me know what my above specific commands are missing? 

Thanks very much!

[1] https://stackoverflow.com/questions/37770467/why-do-i-have-to-git-push-set-upstream-origin-branch
[2] https://github.com/ceph/ceph/pull/38997
[3] https://github.com/openstack-charmers/zaza-openstack-tests/pull/525

ajkavanagh commented:
@zhhuabj The problem is that you did a rebase, which re-writes the commit and loses history (on your own branch/PR).  The rebase, on GitHub related projects, ought to be a "merge" so that the history isn't lost.  This is because GitHub, unlike gerrit, doesn't really make it possible to compare two different rebased commits (version's in a gerrit patch), whereas gerrit makes it very easy.

Thus, once you have reviews in a GitHub PR it's much nicer to the reviewers if you merge upstream changes into your PR so that the previous commits aren't overwritten and reviewing comments still have context.

ajkavanagh commented:
> But for RP, I don't know how to achieve this point. Don't we need to use '--amend', but that will generate many commits, it doesn't seem to be what we want. We may also need '--amend'.

Further, it's desirable for a GitHub PR to have many commits.  It makes it easier on GitHub to see the history of the PR evolving.  Once it is ready for merging, the whole PR can be merged into a single commit, and that is what we typically do do.

AurelienLourot commented:
> zhhuabj force-pushed the zhhuabj:hm-port branch from 6a1863a to b44aa88 6 hours ago

So this is exactly what I wanted to avoid. Now I can't easily diff `6a1863a -> b44aa88` because 6a1863a can't be checked out anymore, and GitHub UI doesn't provide me a way to see this diff. It would really have been nice to have asked before force-pushing, since we explicitly asked not to force-push. I'll try to review now. Anyway, thanks for the contribution.

AurelienLourot commented on zaza/openstack/utilities/generic.py:
```diff
@@ -706,3 +706,23 @@ def attach_file_resource(application_name, resource_name,
+
+
+def get_leaders_and_non_leaders(application_name):
+    """Get leader node and non-leader nodes of percona.
```
Please remove `of percona`

zhhuabj replied:
done

AurelienLourot commented on zaza/openstack/utilities/generic.py:
```diff
@@ -706,3 +706,23 @@ def attach_file_resource(application_name, resource_name,
+def get_leaders_and_non_leaders(application_name):
+    """Get leader node and non-leader nodes of percona.
+
+    Update and set on the object the leader node and list of non-leader
```
Also I think this sentence doesn't apply anymore. Let's remove it, thanks.

zhhuabj replied:
done

AurelienLourot commented on zaza/openstack/utilities/generic.py:
```diff
@@ -706,3 +706,23 @@ def attach_file_resource(application_name, resource_name,
+    Update and set on the object the leader node and list of non-leader
+    nodes.
+
+    :returns: None
```
These two lines about the return value also are wrong. I know it's not your fault, you just moved this code, but let's use this opportunity to get this straight, thanks!

zhhuabj replied:
done

AurelienLourot reviewed (changes requested):
Thanks a lot, you nicely addressed my comments. I have 3 tiny things left, and then this will look really nice. I'm trying to be reactive on this review, to help having this useful change landed quickly. But please don't push-force, as this is really making my life harder, otherwise I'll prioritize other reviews which are easier. Thanks for your understanding, and thanks for the good work!

AurelienLourot reviewed (approved):
LGTM, thanks! To be merged together with https://review.opendev.org/c/openstack/charm-octavia/+/787700

zhhuabj commented:
Self-test results:

1, for non-ovn, it works well -  https://pastebin.ubuntu.com/p/syJqbDh2SR/
2, for ovn, it works well -  https://pastebin.ubuntu.com/p/NcMJW4r5sk/

hemanthnakkina commented:
LGTM, lets wait for CI success @ https://review.opendev.org/c/openstack/charm-octavia/+/787700

AurelienLourot reviewed (changes requested):
Thanks a lot @zhhuabj, as explained in the [charm review](https://review.opendev.org/c/openstack/charm-octavia/+/787700) I think we should disable this new test for now.

If I'm not mistaken I think a disabled test looks like

    def test_mytest(self):
        raise unittest.SkipTest("Skipping because of lp:1951858")
        ...
